### PR TITLE
[ExpressionLanguage] Expr. functions added after first eval/compile didn't exist

### PR DESCRIPTION
**Pull request by maidmaid**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

If we add expr. function after first eval/compile like this:

```php
$el = new ExpressionLanguage();
$el->evaluate('1 + 1');
$el->addFunction(new ExpressionFunction('fn', function () {}, function () {}));
$el->evaluate('fn()');
```
A ``SyntaxError`` is thrown that says ``The function "fn" does not exist around position 1.``. It's the same bug with ``$el->compile('fn()')``.

This PR fixes this.

### CODE REVIEW

**Comment by unkind** [comment]
I think you can just create new instance of `Compiler` in `addFunction` (and in other mutators if any).

**Comment by maidmaid** [comment]
I prefer that ``ExpressionLanguage``, ``Compiler`` and ``Parser`` use the same reference of array of functions. If EL adds a new function, ``Compiler`` and ``Parser`` see this change without be recreated.

File: src/Symfony/Component/ExpressionLanguage/Compiler.php [hunk 1]
```diff
@@ -21,9 +21,9 @@ class Compiler
     private $source;
     private $functions;
 
-    public function __construct(array $functions)
+    public function __construct(array &$functions)
```
Comment (xabbuh) [inline src/Symfony/Component/ExpressionLanguage/Compiler.php]: I am not sure if passing the functions by reference ID actually a good idea as it may lead to unexpected side effects.

Comment (maidmaid) [inline src/Symfony/Component/ExpressionLanguage/Compiler.php]: Is it a break BC?

**Comment by unkind** [comment]
@maidmaid the `ExpressionLanguage` service has a design issue: mutable state. If it would receive functions through constructor only, you didn't encounter it. When you want to fix it by adding more mutability (for `Compiler` and `Parser`), you repeat same mistake.

File: src/Symfony/Component/ExpressionLanguage/Compiler.php (hunk 1 above)
Comment (unkind) [inline src/Symfony/Component/ExpressionLanguage/Compiler.php]: It is.

**Comment by fabpot** [comment]
I would instead forbid changing the ExpressionLanguage functions after calling evaluate the first time.

**Comment by maidmaid** [comment]
How do you see this @fabpot? By throwing an exception?

**Comment by fabpot** [comment]
@maidmaid I think throwing an exception would make sense.

**Comment by maidmaid** [comment]
But throw an exception in existing behavior, is not it a BC break?

**Comment by fabpot** [comment]
No, as it currently does not work anyway.

**Comment by fabpot** [comment]
Any news?

**Comment by maidmaid** [comment]
Sorry for delay... I done changes.

File: src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php [hunk 2]
```diff
@@ -110,10 +110,16 @@ public function parse($expression, $names)
      * @param callable $compiler  A callable able to compile the function
      * @param callable $evaluator A callable able to evaluate the function
      *
+     * @throws \LogicException when register a function after calling evaluate/compilate the first time
+     *
      * @see ExpressionFunction
      */
     public function register($name, $compiler, $evaluator)
     {
+        if ($this->parser) {
```
Comment (xabbuh) [inline src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php]: `null !== $this->parser`

File: src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php [hunk 3]
```diff
@@ -110,10 +110,16 @@ public function parse($expression, $names)
...
+            throw new \LogicException('It\'s impossible to register function after calling evaluate/compilate the first time.');
```
Comment (xabbuh) [inline src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php]: I would reword this a bit to let it sound less confusing:

> Registering functions after calling the evaluate(), compile(), or parse() is not supported.

File: src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php (hunk 2 above)
Comment (maidmaid) [inline src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php]: Done (but why not use automatic boolean casting in fact?).

File: src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php (hunk 3 above)
Comment (maidmaid) [inline src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php]: Done.

File: src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php [hunk 4]
```diff
@@ -110,10 +110,16 @@ public function parse($expression, $names)
      * @param callable $compiler  A callable able to compile the function
      * @param callable $evaluator A callable able to evaluate the function
      *
+     * @throws \LogicException when register a function after calling evaluate(), compile() or parse() the first time
```
Comment (fabpot) [inline src/Symfony/Component/ExpressionLanguage/ExpressionLanguage.php]: I would say `when registering a function after calling evaluate(), compile(), or parse()`.

**Review by fabpot — APPROVED** [review]

**Comment by maidmaid** [comment]
I rebased my PR after update the tests and next :boom:. I'm confused, I must recreate a new PR?

**Comment by maidmaid** [comment]
I rebased from master branch instead 2.7 branch... I have recreated PR #21722, sorry for the noise.
